cpp/src/io/csv/csv_gpu.h

```diff
@@ -23,6 +23,120 @@ namespace cudf {
 namespace io {
 namespace csv {
 namespace gpu {
+
+/**
+ * CSV row parsing context
+ * NONE: No special context (normal parsing)
+ * QUOTE: Within a quoted field
+ * COMMENT: Within a comment line (discard every character until terminator)
+ * EOF: End state (EOF reached)
+ **/
+enum { ROW_CTX_NONE = 0, ROW_CTX_QUOTE = 1, ROW_CTX_COMMENT = 2, ROW_CTX_EOF = 3 };
+
+constexpr uint32_t rowofs_block_dim = 512;
+/// Character block size for gather_row_offsets
+constexpr uint32_t rowofs_block_bytes = rowofs_block_dim * 32;  // 16KB/threadblock
+
+/**
+ * @brief return a row context from a {count, id} pair
+ *
+ * The 32-bit row context consists of the 2-bit parser state stored in the lower 2-bits
+ * and a 30-bit row count in the upper 30 bits.
+ *
+ **/
+inline __host__ __device__ uint32_t make_row_context(uint32_t row_count, uint32_t out_ctx)
+{
+  return row_count * 4 + out_ctx;
```

Comment:
```suggestion
  return row_count << 2 + out_ctx;
```

Reply:
You don't want that (operator precedence).

Reply:
Ah, of course. How about `(row_count << 2) | out_ctx` then?

Reply:
Sure. Either way is identical, so I don't have a strong preference for one way or the other (I'll generally just go for whatever results in the less amount of typing).